### PR TITLE
Fix {}-substitutions failing on container types

### LINKED ISSUE #364: {}-substitutions fail on List[Tuple[float,float]] types
Installing `breifast` branch [dynspec](https://github.com/ratt-ru/breifast/tree/dynspec) and trying to produce dynamic spectra using the recipe `breifast/breifast-tron/breifast/recipes/mkdynspec-recipe.yml` like this:

```
stimela run breifast.recipes::mkdynspec-recipe.yml ms=/net/nash/home/ianh/Globular/OmegaCen/1541903773_sdp_l0_1024ch
_OmegaCen.ms data=CORRECTED_DATA WeightCol=WEIGHT_SPECTRUM model=MODEL_DATA srclist=/home/myburgh/DynamicSpectra/OmegaCenMSP/omegacen_msp.txt rad=1 noff=3
0 DoTar=0 NCPU=32 OutDirName=/home/myburgh/DynamicSpectra/OmegaCenMSP/1541903773_sdp_l0_1024ch_OmegaCen_DynSpecs_1541903773_beamModel CacheDir=/home/mybur
gh DDFParset=/home/myburgh/im-HR1.parset kernel=[[30,120],[90,360]] std-scale=2.5 debug=True stokes=I cmap=bwr
```
Yields errors:

Image

Then installing `inspect_dynspec` as:
```
 git clone https://github.com/ratt-ru/fatalii.git && \
 cd fatalii && \
 git checkout Dynspec_no_lc && \
 python -m pip install --no-cache-dir ./inspect_dynspec
```

And running the `inspect_dynspec` cab directly:

```
stimela run breifast.cabs::inspect_dynspec.yml root=. output=~/NormAndSmoothPaperSubtract kernel=[[30,515],[35,600]] std-scale=3.0 cmap=bwr plot-for-paper=True debug=True stokes=I
```
validates the arguments just fine:

Image

Is it possible that in the recipe instance, parameters are being altered due to the [format](https://github.com/ratt-ru/breifast/blob/dynspec/breifast-tron/breifast/recipes/mkdynspec-recipe.yml#L70) specifications, and are then no longer suitable for the cab to injest?

### PR DESCRIPTION
**Pull request by gijzelaerr**
## Summary
- When a `{}`-substitution converts a container value to its string representation (e.g. `[(30.0, 120.0), (90.0, 360.0)]`), pydantic cannot parse it back into the expected type like `List[Tuple[float,float]]`
- Adds a pre-validation step in `validate_parameters` that detects string values for container dtypes (`list`, `tuple`, `dict`) and parses them via `ast.literal_eval` (with `yaml.safe_load` as fallback) before passing to pydantic
- Adds 10 tests covering the core issue case, various container types, edge cases, and error handling

Fixes caracal-pipeline/stimela#364

## Test plan
- [x] All 77 existing + new tests pass
- [x] Core case `List[Tuple[float,float]]` from string validated correctly
- [x] `List[int]`, `List[float]`, `Dict[str,int]`, `Tuple[int,float]` all work from string
- [x] Malformed strings still raise `ParameterValidationError`
- [x] Scalar dtypes are unaffected
- [x] Ruff lint and format pass

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: scabha/validate.py [hunk 1]
```diff
@@ -209,6 +210,21 @@ def validate_parameters(
...
+                # yaml.safe_load is tried as a fallback for simpler list/dict
+                # notations.
+                for parser in (ast.literal_eval, yaml.safe_load):
+                    try:
+                        parsed = parser(value)
+                    except Exception:
+                        continue
+                    if isinstance(parsed, (list, tuple, dict)):
+                        inputs[name] = parsed
+                        break
```
[the remark this answers is not in the document]
Comment (gijzelaerr) [reply scabha/validate.py]: Fixed in f61c946. Added a `_is_container_dtype()` helper that unwraps `Optional[...]`/`Union[...]` types before checking if the inner type's origin is `list`, `tuple`, or `dict`. It also handles bare builtins (e.g. `dtype: list` where `get_origin(list)` is `None`) by checking `dtype in (list, tuple, dict)` directly. Added a test for `Optional[List[int]]` to cover this case.

File: tests/test_validate.py [hunk 2]
```diff
@@ -194,6 +194,92 @@ def test_nested_rejects_4d_basis(self):
...
+        )
+        assert out["x"] == 42
+
+    def test_yaml_fallback_for_simple_list(self):
+        """YAML notation like [1, 2, 3] should also work via yaml.safe_load."""
+        out = run_validate(
+            {"x": "[1, 2, 3]"},
+            {"x": make_schema("List[int]")},
+        )
+        assert out["x"] == [1, 2, 3]
```
[the remark this answers is not in the document]
Comment (gijzelaerr) [reply tests/test_validate.py:280]: Fixed in f61c946. Changed the test input from `"[1, 2, 3]"` (which `ast.literal_eval` handles fine) to YAML block-list syntax `"- 1\n- 2\n- 3"`, which `ast.literal_eval` rejects. This ensures the test actually exercises the `yaml.safe_load` fallback path.

**Review by o-smirnov — APPROVED** [review]
LGTM
